## Visual Assessment: Acquaint Yourself
This Auralin Phase II clinical trial dataset comes in three tables: `patients`, `treatments`, and `adverse_reactions`. Acquaint yourself with them through visual assessment below.

### Gather

In [1]:
!git clone https://github.com/CherifArsanious/assessing-cleaning-data.git

Cloning into 'assessing-cleaning-data'...
remote: Enumerating objects: 15, done.
remote: Counting objects: 100% (15/15), done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 15 (delta 5), reused 0 (delta 0), pack-reused 0
Unpacking objects: 100% (15/15), done.


In [0]:
import pandas as pd

In [0]:
patients = pd.read_csv('assessing-cleaning-data/patients.csv')
treatments = pd.read_csv('assessing-cleaning-data/treatments.csv')
adverse_reactions = pd.read_csv('assessing-cleaning-data/adverse_reactions.csv')

### Assess
In the cells below, each column of each table in this clinical trial dataset is described. To see the table that goes hand in hand with these descriptions, display each table in its entirety by displaying the pandas DataFrame that it was gathered into. This task is the mechanical part of visual assessment in pandas.

In [4]:
# Display the patients table
patients

patient_id assigned_sex given_name  ... weight height   bmi
0             1       female        Zoe  ...  121.7     66  19.6
1             2       female     Pamela  ...  118.8     66  19.2
2             3         male        Jae  ...  177.8     71  24.8
3             4         male       Liêm  ...  220.9     70  31.7
4             5         male        Tim  ...  192.3     27  26.1
..          ...          ...        ...  ...    ...    ...   ...
498         499         male    Mustafa  ...  181.1     72  24.6
499         500         male      Ruman  ...  239.6     70  34.4
500         501       female      Jinke  ...  171.2     67  26.8
501         502       female    Chidalu  ...  176.9     67  27.7
502         503         male        Pat  ...  138.2     71  19.3

[503 rows x 14 columns]

`patients` columns:
- **patient_id**: the unique identifier for each patient in the [Master Patient Index](https://en.wikipedia.org/wiki/Enterprise_master_patient_index) (i.e. patient database) of the pharmaceutical company that is producing Auralin
- **assigned_sex**: the assigned sex of each patient at birth (male or female)
- **given_name**: the given name (i.e. first name) of each patient
- **surname**: the surname (i.e. last name) of each patient
- **address**: the main address for each patient
- **city**: the corresponding city for the main address of each patient
- **state**: the corresponding state for the main address of each patient
- **zip_code**: the corresponding zip code for the main address of each patient
- **country**: the corresponding country for the main address of each patient (all United states for this clinical trial)
- **contact**: phone number and email information for each patient
- **birthdate**: the date of birth of each patient (month/day/year). The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is  age >= 18 *(there is no maximum age because diabetes is a [growing problem](http://www.diabetes.co.uk/diabetes-and-the-elderly.html) among the elderly population)*
- **weight**: the weight of each patient in pounds (lbs)
- **height**: the height of each patient in inches (in)
- **bmi**: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m<sup>2</sup> where kg is a person's weight in kilograms and m<sup>2</sup> is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. *The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is 16 >= BMI >= 38.*

In [5]:
patients.head()

patient_id assigned_sex given_name  surname  ...  birthdate weight height   bmi
0           1       female        Zoe  Wellish  ...  7/10/1976  121.7     66  19.6
1           2       female     Pamela     Hill  ...   4/3/1967  118.8     66  19.2
2           3         male        Jae   Debord  ...  2/19/1980  177.8     71  24.8
3           4         male       Liêm     Phan  ...  7/26/1951  220.9     70  31.7
4           5         male        Tim  Neudorf  ...  2/18/1928  192.3     27  26.1

[5 rows x 14 columns]

In [6]:
patients.tail()

patient_id assigned_sex given_name  ... weight height   bmi
498         499         male    Mustafa  ...  181.1     72  24.6
499         500         male      Ruman  ...  239.6     70  34.4
500         501       female      Jinke  ...  171.2     67  26.8
501         502       female    Chidalu  ...  176.9     67  27.7
502         503         male        Pat  ...  138.2     71  19.3

[5 rows x 14 columns]

In [7]:
patients.sample()

patient_id assigned_sex given_name  ... weight height   bmi
491         492       female    Vanessa  ...  149.8     67  23.5

[1 rows x 14 columns]

In [8]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [9]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [10]:
# Display the treatments table
treatments

given_name       surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0     veronika      jindrová  41u - 48u  ...        7.63       7.20           NaN
1       elliot    richardson          -  ...        7.56       7.09          0.97
2     yukitaka      takenaka          -  ...        7.68       7.25           NaN
3         skye    gormanston  33u - 36u  ...        7.97       7.62          0.35
4       alissa        montez          -  ...        7.78       7.46          0.32
..         ...           ...        ...  ...         ...        ...           ...
275     albina      zetticci  45u - 51u  ...        7.93       7.73          0.20
276       john   teichelmann          -  ...        7.90       7.58           NaN
277     mathea       lillebø  23u - 36u  ...        9.04       8.67          0.37
278     vallie        prince  31u - 38u  ...        7.64       7.28          0.36
279     samúel  guðbrandsson  53u - 56u  ...        8.00       7.64          0.36

[280 rows x 7 columns]

In [11]:
patients.query('city=="New York"').shape[0]

18

In [12]:
patients.query('surname=="Doe"').shape[0]

6

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before.  All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

`treatments` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial
- **auralin**: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) *and* the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the [international unit](https://en.wikipedia.org/wiki/International_unit) of measurement and the standard measurement for insulin.
- **novodra**: same as above, except for patients that continued treatment with Novodra
- **hba1c_start**: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The [HbA1c test](https://depts.washington.edu/uwcoe/healthtopics/diabetes.html) measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- **hba1c_end**: the patient's HbA1c level at the end of the last week of treatment
- **hba1c_change**: the change in the patient's HbA1c level from the start of treatment to the end, i.e., `hba1c_start` - `hba1c_end`. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [13]:
# Display the adverse_reactions table
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

`adverse_reactions` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **adverse_reaction**: the adverse reaction reported by the patient

Additional useful information:
- [Insulin resistance varies person to person](http://www.tudiabetes.org/forum/t/how-much-insulin-is-too-much-on-a-daily-basis/9804/5), which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This [diversity](https://www.clinicalleader.com/doc/an-fda-perspective-on-patient-diversity-in-clinical-trials-0001) is reflected in the `patients` table.
- Ensuring column names are descriptive enough is an important step in acquainting yourself with the data. 'Descriptive enough' is subjective. Ideally you want short column names (so they are easier to type and read in code form) but also fully descriptive. Length vs. descriptiveness is a tradeoff and common debate (a [similar debate](https://softwareengineering.stackexchange.com/questions/176582/is-there-an-excuse-for-short-variable-names) exists for variable names). The *auralin* and *novodra* column names are probably not descriptive enough, but you'll address that later so don't worry about that for now.

In [14]:
adverse_reactions['adverse_reaction'].value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
throat irritation             2
nausea                        2
cough                         2
Name: adverse_reaction, dtype: int64

## Cleaning Data

In [0]:
# Making copies of our dataframes 
patients_clean=patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

## treatments: Missing records (280 instead of 350)
### Define
merge the two tables the one of treatments and the other one treatments_cut that holds the rest of records

In [0]:
df_cut=pd.read_csv('assessing-cleaning-data/treatments_cut.csv')

In [17]:
df_cut.head()

given_name    surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0      jožka  resanovič  22u - 30u  ...        7.56       7.22          0.34
1  inunnguaq   heilmann  57u - 67u  ...        7.85       7.45           NaN
2      alwin   svensson  36u - 39u  ...        7.78       7.34           NaN
3        thể      lương          -  ...        7.64       7.22          0.92
4     amanda    ribeiro  36u - 44u  ...        7.85       7.47          0.38

[5 rows x 7 columns]

In [18]:
df_cut.shape[0]

70

In [19]:
treatments_clean.shape[0]

280

In [0]:
treatments_clean=pd.concat([treatments_clean,df_cut],axis=0,ignore_index=True)

In [21]:
treatments_clean.index.duplicated()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [22]:
treatments_clean.shape[0]

350

### treatments: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)
Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one Define, Code, and Test header occurs multiple times in this notebook.

In [23]:
treatments_clean.head()

given_name     surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0   veronika    jindrová  41u - 48u  ...        7.63       7.20           NaN
1     elliot  richardson          -  ...        7.56       7.09          0.97
2   yukitaka    takenaka          -  ...        7.68       7.25           NaN
3       skye  gormanston  33u - 36u  ...        7.97       7.62          0.35
4     alissa      montez          -  ...        7.78       7.46          0.32

[5 rows x 7 columns]

In [0]:
treatments_clean['hba1c_change']=treatments_clean['hba1c_start']-treatments_clean['hba1c_end']

In [25]:
treatments_clean['hba1c_change'].isnull().sum()

0

## Contact column in patients table contains two variables: phone number and email
Extract the phone number and email variables from the contact column using regular expressions and pandas' str.extract method. Drop the contact column when done.

In [0]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

In [27]:
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [28]:
patients_clean.phone_number.sample(25)

304         502-902-6188
313         251-359-2088
287         619 818 3436
369         619-652-0285
475         650-962-7179
274         740-552-2444
381    +1 (870) 457-5086
419         317-956-6166
156    +1 (612) 208-2965
357         937-592-9410
322         254-546-2728
108         620 940 1131
460    +1 (510) 238-1151
446         601-418-0102
203         636-442-6946
70          740 629 5172
384    +1 (605) 440-5492
74          918-975-7594
339         952 431 5166
77          859-297-3368
407         405-470-5088
166         903-749-8046
310         913 322 9114
222         502-367-6167
355         401-456-2797
Name: phone_number, dtype: object

In [29]:
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

## Define
Melt the auralin and novodra columns to a treatment and a dose column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain start_dose and end_dose columns. Drop the intermediate dose column.

In [30]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean


given_name     surname  hba1c_start  ...  hba1c_change  treatment       dose
0     veronika    jindrová         7.63  ...          0.43    auralin  41u - 48u
1       elliot  richardson         7.56  ...          0.47    auralin          -
2     yukitaka    takenaka         7.68  ...          0.43    auralin          -
3         skye  gormanston         7.97  ...          0.35    auralin  33u - 36u
4       alissa      montez         7.78  ...          0.32    auralin          -
..         ...         ...          ...  ...           ...        ...        ...
695     rovzan     kishiev         7.75  ...          0.34    novodra          -
696      jakob    jakobsen         7.96  ...          0.45    novodra  28u - 26u
697      bernd   schneider         7.74  ...          0.30    novodra          -
698      berta  napolitani         7.68  ...          0.47    novodra  42u - 44u
699     armina       sauvé         7.86  ...          0.46    novodra          -

[700 rows x 7 columns]

In [31]:
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean.head()

given_name     surname  hba1c_start  ...  hba1c_change  treatment       dose
0   veronika    jindrová         7.63  ...          0.43    auralin  41u - 48u
3       skye  gormanston         7.97  ...          0.35    auralin  33u - 36u
6     sophia      haugen         7.65  ...          0.38    auralin  37u - 42u
7      eddie      archer         7.89  ...          0.34    auralin  31u - 38u
9       asia     woźniak         7.76  ...          0.39    auralin  30u - 36u

[5 rows x 7 columns]

In [32]:
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean=treatments_clean.drop('dose',axis=1)
treatments_clean.head()

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


given_name     surname  hba1c_start  ...  treatment  dose_start dose_end
0   veronika    jindrová         7.63  ...    auralin         41u      48u
3       skye  gormanston         7.97  ...    auralin         33u      36u
6     sophia      haugen         7.65  ...    auralin         37u      42u
7      eddie      archer         7.89  ...    auralin         31u      38u
9       asia     woźniak         7.76  ...    auralin         30u      36u

[5 rows x 8 columns]

## Adverse reaction should be part of the treatments table
### Define
Merge the adverse_reaction column to the treatments table, joining on given name and surname.

In [33]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean,
                            on=['given_name', 'surname'], how='left')
#adding argument how='left' to preserve all the rows of the treatments 
#table (the left table)as not all patients suffered from adverse effects so a normal 
#merge without the how left parameter would have 
#merged only the common rows between the two tables
treatments_clean.head(25)

given_name       surname  ...  dose_end   adverse_reaction
0      veronika      jindrová  ...       48u                NaN
1          skye    gormanston  ...       36u                NaN
2        sophia        haugen  ...       42u                NaN
3         eddie        archer  ...       38u                NaN
4          asia       woźniak  ...       36u                NaN
5        joseph           day  ...       36u       hypoglycemia
6       roxanne     andreyeva  ...       38u                NaN
7        simone  baumgaertner  ...       37u                NaN
8          enco        žibrik  ...       68u                NaN
9       camilla      zaitseva  ...       37u                NaN
10        tekla       walczak  ...       39u                NaN
11  brancaleone         russo  ...       60u                NaN
12         isac          berg  ...       41u                NaN
13      clinton        miller  ...       51u  throat irritation
14       eugene       mironov  ...       49u                NaN
15  szilveszter         totth  ...       39u                NaN
16    alexander     mathiesen  ...       58u                NaN
17          chỉ           lâm  ...       48u                NaN
18    władysław     wieczorek  ...       37u                NaN
19     kristján       ingason  ...       55u                NaN
20       marija      grubišić  ...       43u                NaN
21        sauli    koivuniemi  ...       47u                NaN
22      mariana         souza  ...       42u                NaN
23   kristoffer     martinsen  ...       37u                NaN
24           mỹ         quynh  ...       64u                NaN

[25 rows x 9 columns]

In [34]:
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

/usr/local/lib/python3.6/dist-packages/pandas/core/generic.py:5208: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [35]:
treatments_clean

hba1c_start  hba1c_end  ...           adverse_reaction patient_id
0           7.63       7.20  ...                        NaN        225
1           7.97       7.62  ...                        NaN        242
2           7.65       7.27  ...                        NaN        345
3           7.89       7.55  ...                        NaN        276
4           7.76       7.37  ...                        NaN         15
..           ...        ...  ...                        ...        ...
344         7.51       7.06  ...                     nausea        153
345         7.67       7.30  ...                        NaN        420
346         9.21       8.80  ...  injection site discomfort        336
347         7.96       7.51  ...               hypoglycemia         25
348         7.68       7.21  ...  injection site discomfort        477

[349 rows x 8 columns]

In [36]:
list(treatments_clean)

['hba1c_start',
 'hba1c_end',
 'hba1c_change',
 'treatment',
 'dose_start',
 'dose_end',
 'adverse_reaction',
 'patient_id']

In [0]:
treatments_clean=treatments_clean[['patient_id','treatment',
 'dose_start',
 'dose_end',
 'adverse_reaction',
 'hba1c_start',
 'hba1c_end',
 'hba1c_change',
 ]]

In [38]:
treatments_clean

patient_id treatment dose_start  ... hba1c_start hba1c_end  hba1c_change
0           225   auralin        41u  ...        7.63      7.20          0.43
1           242   auralin        33u  ...        7.97      7.62          0.35
2           345   auralin        37u  ...        7.65      7.27          0.38
3           276   auralin        31u  ...        7.89      7.55          0.34
4            15   auralin        30u  ...        7.76      7.37          0.39
..          ...       ...        ...  ...         ...       ...           ...
344         153   novodra        55u  ...        7.51      7.06          0.45
345         420   novodra        26u  ...        7.67      7.30          0.37
346         336   novodra        22u  ...        9.21      8.80          0.41
347          25   novodra        28u  ...        7.96      7.51          0.45
348         477   novodra        42u  ...        7.68      7.21          0.47

[349 rows x 8 columns]

In [76]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        503 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: datetime64[ns](1), float64(2), int64(2), object(10)
memory usage: 59.1+ KB


In [85]:
patients_clean['zip_code']

0      92390
1      61812
2      68467
3      07095
4      36303
       ...  
498    03852
499    86341
500    64110
501      NaN
502      NaN
Name: zip_code, Length: 503, dtype: object

In [0]:
patients_clean['zip_code']=patients_clean['zip_code'].astype(str).str[:-2].str.pad(5,fillchar='0')


In [0]:
import numpy as np
patients_clean['zip_code']=patients_clean['zip_code'].replace('0000n',np.nan)

In [0]:
patients_clean['height']=patients_clean['height'].replace({27:72})

In [56]:
patients_clean.query('patient_id==5')

patient_id assigned_sex given_name  ...   bmi  phone_number                email
4           5         male        Tim  ...  26.1  334-515-7487  TimNeudorf@cuvox.de

[1 rows x 15 columns]

In [60]:
patients_clean['given_name'].duplicated().all()

False

In [0]:
patients_clean['given_name']=patients_clean['given_name'].replace({'Dsvid':'David'})

In [0]:
patients_clean['birthdate']=pd.to_datetime(patients_clean['birthdate'])

In [75]:
patients_clean['birthdate']

0     1976-07-10
1     1967-04-03
2     1980-02-19
3     1951-07-26
4     1928-02-18
         ...    
498   1959-04-10
499   1948-03-26
500   1971-01-13
501   1952-02-13
502   1954-05-03
Name: birthdate, Length: 503, dtype: datetime64[ns]

In [77]:
treatments_clean.head(10)

patient_id treatment dose_start  ... hba1c_start hba1c_end  hba1c_change
0         225   auralin        41u  ...        7.63      7.20          0.43
1         242   auralin        33u  ...        7.97      7.62          0.35
2         345   auralin        37u  ...        7.65      7.27          0.38
3         276   auralin        31u  ...        7.89      7.55          0.34
4          15   auralin        30u  ...        7.76      7.37          0.39
5          70   auralin        29u  ...        7.70      7.19          0.51
6          70   auralin        29u  ...        7.70      7.19          0.51
7          18   auralin        29u  ...        9.54      9.14          0.40
8         424   auralin        27u  ...        7.74      7.30          0.44
9         292   auralin        55u  ...        7.78      7.34          0.44

[10 rows x 8 columns]

In [0]:
treatments_clean['dose_start']=treatments_clean['dose_start'].str[:-1].astype('int')

In [0]:
treatments_clean['dose_end']=treatments_clean['dose_end'].str[:-1].astype('int')

In [80]:
treatments_clean[['dose_start','dose_end']]

dose_start  dose_end
0            41        48
1            33        36
2            37        42
3            31        38
4            30        36
..          ...       ...
344          55        51
345          26        23
346          22        23
347          28        26
348          42        44

[349 rows x 2 columns]

In [81]:
treatments_clean[['dose_start','dose_end']].dtypes

dose_start    int64
dose_end      int64
dtype: object

In [0]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

In [87]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
OK    13
MI    13
LA    13
NJ    12
VA    11
MS    10
WI    10
MN     9
AL     9
TN     9
IN     9
WA     8
NC     8
KY     8
MO     7
NV     6
KS     6
ID     6
NE     6
CT     5
IA     5
SC     5
CO     4
AR     4
AZ     4
RI     4
ND     4
ME     4
SD     3
DE     3
OR     3
WV     3
MD     3
MT     2
VT     2
DC     2
NM     1
NH     1
WY     1
AK     1
Name: state, dtype: int64

In [0]:
patients_clean.assigned_sex=patients_clean.assigned_sex.astype('category')
patients_clean.state=patients_clean.state.astype('category')

In [89]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        489 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [0]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

In [91]:
patients_clean.phone_number.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

In [0]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

In [93]:
# Should be no Doe records
patients_clean.surname.value_counts()

Jakobsen      3
Taylor        3
Johnson       2
Tucker        2
Schiavone     2
             ..
Hsu           1
Wiśniewska    1
Koldenhof     1
Azevedo       1
Gunnarsson    1
Name: surname, Length: 465, dtype: int64

In [94]:
patients_clean.query('surname=="Jakobsen"')

patient_id assigned_sex  ... phone_number                         email
24           25         male  ...  18458587707     JakobCJakobsen@einrot.com
29           30         male  ...  18458587707     JakobCJakobsen@einrot.com
432         433       female  ...  19792030438  KarenJakobsen@jourrapide.com

[3 rows x 15 columns]

In [0]:
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

In [0]:
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = patients_clean.loc[mask, column_name] * 2.20462